server/src/main/java/org/cloudfoundry/identity/uaa/zone/MultitenantVmidentityClientDetailsService.java

```diff
@@ -0,0 +1,335 @@
+        } catch (NoSuchOIDCClientException e) {
+            logger.error("No client found with id: " + clientId + " in identity zone " + IdentityZoneHolder.get().getName(), e);
+            throw new NoSuchClientException("No client found with id = " + clientId + " in identity zone " + IdentityZoneHolder.get().getName());
+        } catch (Exception e) {
+            logger.error("Unable to add client secret with id: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to add client secret");
+        }
+    }
+
+    @Override
+    public void deleteClientSecret(String clientId) throws NoSuchClientException {
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            ClientDetails clientDetails = loadClientByClientId(clientId);
+            String clientSecret = clientDetails.getClientSecret().split(" ")[1];
+            OIDCClient client = mapClientDetails(clientDetails, clientSecret);
+            idmClient.setOIDCClient(tenant, client);
+        } catch (NoSuchOIDCClientException e) {
+            logger.error("No client found with id: " + clientId + " in identity zone " + IdentityZoneHolder.get().getName(), e);
+            throw new NoSuchClientException("No client found with id = " + clientId + " in identity zone " + IdentityZoneHolder.get().getName());
+        } catch (Exception e) {
+            logger.error("Unable t update client secret with id: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to update client secret");
+        }
+    }
+
+    private static OIDCClient mapClientDetails(ClientDetails details, String clientSecret) {
+        OIDCClient.Builder builder = new OIDCClient.Builder(details.getClientId());
+        // It would be better to treat these as 'Collection' types instead...
+        builder.redirectUris(convertToList(details.getRegisteredRedirectUri()));
+        builder.resourceIds(convertToList(details.getResourceIds()));
+        builder.scopes(convertToList(details.getScope()));
+        builder.authorizedGrantTypes(convertToList(details.getAuthorizedGrantTypes()));
+        builder.authorities(authoritiesToListOfStrings(details.getAuthorities()));
+
+        // builder.accessTokenValiditySeconds(details.getAccessTokenValiditySeconds()); TODO access token validity
+        // builder.refreshTokenValiditySeconds(details.getRefreshTokenValiditySeconds()); TODO refresh token validity
+
+        if (clientSecret != null) {
+            builder.clientSecret(clientSecret);
+        }
+
+        Set<String> autoApproveScopes = new HashSet<String>();
+        Map<String, Object> additionalInformation = details.getAdditionalInformation();
+        Object autoApprovedFromAddInfo = additionalInformation.get(ClientConstants.AUTO_APPROVE);
+        if (autoApprovedFromAddInfo != null) {
+           if ((autoApprovedFromAddInfo instanceof Boolean && (Boolean) autoApprovedFromAddInfo || "true".equals(autoApprovedFromAddInfo))) {
```

Comment:
we have this same logic some other place. probably makes sense to make a function...